server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,297 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.common.Change;
+import com.linecorp.centraldogma.common.Entry;
+import com.linecorp.centraldogma.common.Revision;
+import com.linecorp.centraldogma.server.command.CommitResult;
+import com.linecorp.centraldogma.server.internal.storage.repository.MirrorConfig;
+import com.linecorp.centraldogma.server.internal.storage.repository.RepositoryMetadataException;
+import com.linecorp.centraldogma.server.mirror.MirrorCredential;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+class MirroringMigrationService {
+
+    private static final Logger logger = LoggerFactory.getLogger(MirroringMigrationService.class);
+
+    @VisibleForTesting
+    static final String PATH_LEGACY_MIRRORS = "/mirrors.json";
+    @VisibleForTesting
+    static final String PATH_LEGACY_CREDENTIALS = "/credentials.json";
+    private static final String PATH_LEGACY_MIRRORS_BACKUP = PATH_LEGACY_MIRRORS + ".bak";
+    private static final String PATH_LEGACY_CREDENTIALS_BACKUP = PATH_LEGACY_CREDENTIALS + ".bak";
+
+    private final ProjectManager projectManager;
+
+    @Nullable
+    private List<String> shortWords;
+
+    MirroringMigrationService(ProjectManager projectManager) {
+        this.projectManager = projectManager;
+    }
+
+    void migrate() throws Exception {
+        for (Project project : projectManager.list().values()) {
+            logger.info("Migrating mirrors and credentials in the project: {} ...", project.name());
+            boolean processed = false;
+            processed |= migrateCredentials(project);
+            // Update the credential IDs in the mirrors.json file.
+            processed |= migrateMirrors(project);
+            if (processed) {
+                logger.info("Mirrors and credentials in the project: {} have been migrated.", project.name());
+            } else {
+                logger.info("No legacy configurations of mirrors and credentials found in the project: {}.",
+                            project.name());
+            }
+
+        }
+        shortWords = null;
+    }
+
+    private boolean migrateMirrors(Project project) throws Exception {
+
+        final ArrayNode mirrors = getMetaData(project, PATH_LEGACY_MIRRORS);
+        if (mirrors == null) {
+            return false;
+        }
+
+        // Back up the old mirrors.json file and don't use it anymore.
+        rename(project, PATH_LEGACY_MIRRORS, PATH_LEGACY_MIRRORS_BACKUP);
+
+        final List<MirrorCredential> credentials = project.metaRepo().credentials()
```

Comment:
Question) Say that the `mirrors.json` file is migrated, but this line throws an exception.
Do we manually revert the file before trying again?

Would it make sense to rename the file at the last step instead?

In the interest of making `migrateMirrors` idempotent, what do you think of assigning a deterministic mirrorId as well? (e.g. using a hash)

Reply:
Implemented rollback logic that removes all new migrated files and reverts the backup files.